Tail: Implement -F, --retry (Fixes #847)

GaurangTandon opened the pull request:
- [x] Added options for -F, --retry
- [x] Added platform specific link counting
- [x] Actually implemented --retry
- [ ] Implement tests for --retry, -F

sylvestre commented:
merci!

sylvestre commented on src/uu/tail/src/platform/unix.rs:
```diff
@@ -40,3 +42,12 @@ pub fn supports_pid_checks(pid: self::Pid) -> bool {
 }
+
+#[inline]
+pub fn links_count(f: &File) -> u64 {
```
is that expected?

sylvestre commented on src/uu/tail/src/platform/windows.rs:
```diff
@@ -57,3 +57,13 @@ impl Drop for ProcessChecker {
 }
+
+#[inline]
+pub fn links_count(f: &File) -> u64 {
```
please fix that :)

GaurangTandon replied:
I am not sure what is to be fixed here? We are going to need the `links_count` function to be able to detect if an open File object got deleted while we're reading it. I moved it to `platform/` because it's a platform dependent function.

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -110,14 +118,29 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
     }
 
-    settings.follow = given_options.opt_present("f");
+    if given_options.opt_present("F") {
```
if you are motivated, you could move this to clap (in a different PR). It would simplify the argument management

GaurangTandon replied:
Sure, I'll do that after this is finished.

GaurangTandon commented:
Sorry for the force push, I had to fix the commit message.

GaurangTandon commented:
I think I'm done with the implementation :D Two questions:

1. Is there some documentation on the test macros so I can write a couple for my implementation?
2. Is there a show_info! alternative to show_warning!/show_error!?

sylvestre commented:
> I have done `cargo +1.32.0 update`, but even then some builds are failing, I am not sure why?

The Ubuntu one is an intermittent issue on the CI "test_chown::test_passgrp::test_uid2usr "
Windows is a bug in your code (File doesn't seem to be imported)
https://github.com/uutils/coreutils/pull/1682/checks?check_run_id=1667925830
```

error[E0412]: cannot find type `File` in this scope
Error:   --> src\uu\tail\src\platform\windows.rs:62:24
   |
62 | pub fn links_count(f: &File) -> u64 {
   |                        ^^^^ not found in this scope
```

For the MinRustV
```
Caused by:
  invalid serialized PackageId for key `package.dependencies`
Error: Incompatible 'Cargo.lock' format; try `cargo +1.32.0 update`
```
it means that you didn't run the command (or that, if you use an IDE, it regenerated it using a more recent version of Rust, I agree it is a pain)

> Is there some documentation on the test macros so I can write a couple of tests for my implementation?

The best way is to look how it is done elsewhere :)
https://github.com/uutils/coreutils/tree/master/tests/by-util

> Is there a show_info! alternative to show_warning!/show_error!?

https://github.com/uutils/coreutils/blob/d00356eab0ea98bab7ed5b0aa1e262673f3a9e3c/src/uucore/src/lib/macros.rs#L38 ?

GaurangTandon commented:
Thanks! I have added the windows fix and used show_info, not sure why I couldn't find it earlier. Questions:

1. I tried `rustup default 1.32.0` as well as `cargo +1.32.0 build` but both of them fail the builds due to E0658. If I run `cargo +1.32.0 update` instead of them then the MinV test has failed once again. I don't use an IDE, so I am sure the lockfile wasn't regenerated. I am confused :( 

2. I added one test however it gets timed out. [`f84f6e8` (#1682)](https://github.com/uutils/coreutils/pull/1682/commits/f84f6e8a7ce2566e2134629e078084a08bdf5417)?file-filters%5B%5D=.rs#diff-149eaf1289c45326ed003b93227cacb5a8957ab31a596e9f5a831010528b7562R98 See the [Travis build](https://travis-ci.org/github/uutils/coreutils/jobs/753639875). I found that `read_size` gets stuck. Inside it, `.read(output.as_mut_slice())` is the culprit. I am not sure how to fix it though, what may be causing the problem? :( My test function is almost identical to the others used for testing `follow`.

GaurangTandon commented:
@sylvestre I have marked the test I added as ignored. The general idea of the test is as follows:

1. Run command `tail --retry -f file.txt`, but this file.txt does not exist yet.
2. Expect a message from tail
3. Create the file
4. Expect another message
5. Write content to file.
6. Expect a last message.

I have coded exactly this into the test, yet it gets stuck for more than a minute at step 2. It would be great if you could look into the test once to see where the problem is. Thanks! ^_^

sylvestre commented on src/uu/tail/src/platform/windows.rs:
```diff
@@ -57,3 +58,13 @@ impl Drop for ProcessChecker {
     true
 }
+
+#[inline]
+pub fn links_count(f: &File) -> u64 {
```
please fix the warning

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -214,39 +243,51 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
 
         for filename in &files {
+            let path = Path::new(filename);
+            if path.is_dir() {
```
please add some comments explaining the why here
maybe won't be needed if file_objs is named correctly

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -214,39 +243,51 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
         let multiple = files.len() > 1;
         let mut first_header = true;
-        let mut readers = Vec::new();
+        let mut file_objs = Vec::new();
```
file_objs ? I would prefer you use an explicit variable name. 
Not variable name explaining the type

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -214,39 +243,51 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
+                continue;
+            }
+
+            let file = file.unwrap();
```
Don't we have to check that the open operation worked?

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -214,39 +243,51 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
+            file_objs.push(Some(file));
+        }
+
+        for (i, file) in file_objs.iter_mut().enumerate() {
```
please add a comment explaining why this loop

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -214,39 +243,51 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
+        }
+
+        for (i, file) in file_objs.iter_mut().enumerate() {
+            if file.is_none() {
```
ditto here

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -345,22 +386,83 @@ pub fn parse_size(mut size_slice: &str) -> Result<u64, ParseSizeErr> {
+fn follow_with_retry(file_objs: &mut [Option<File>], filenames: &[String], settings: &Settings) {
+    assert!(settings.follow || settings.retry);
+
+    let total = filenames.len();
```
please rename it for something more explicit like

number_of_files

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -345,22 +386,83 @@ pub fn parse_size(mut size_slice: &str) -> Result<u64, ParseSizeErr> {
+    assert!(settings.follow || settings.retry);
+
+    let total = filenames.len();
+    let mut last = total - 1;
```
last_file ?

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -345,22 +386,83 @@ pub fn parse_size(mut size_slice: &str) -> Result<u64, ParseSizeErr> {
     let mut process = platform::ProcessChecker::new(settings.pid);
+    let mut open_file_count = total;
+
+    for file in file_objs.iter() {
```
please add comment

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -345,22 +386,83 @@ pub fn parse_size(mut size_slice: &str) -> Result<u64, ParseSizeErr> {
         read_some = false;
 
-        for (i, reader) in readers.iter_mut().enumerate() {
+        for i in 0..total {
```
please add comment

sylvestre commented on src/uu/tail/src/tail.rs:
```diff
@@ -345,22 +386,83 @@ pub fn parse_size(mut size_slice: &str) -> Result<u64, ParseSizeErr> {
+                }
+            }
+
+            let file = file_objs[i].as_mut().unwrap();
```
maybe check errors here?

sylvestre commented:
I won't have time to try to fix the test right now.
Please have a second look. Look at other examples

sylvestre commented:
@GaurangTandon 
are you planning to finish it?
thanks

GaurangTandon commented:
Hello @sylvestre I would love to finish it, unfortunately, I do not have any time upto at least the next month :(

If you want to discuss why I had made what kinds of changes, I'm there on the uutils Discord server already (`sigma_g`) so feel free to discuss :D

sylvestre commented:
@GaurangTandon are you still planning to finish it? thanks

GaurangTandon commented:
Sorry, it doesn't look like I'll be able to complete this anytime soon. I haven't had time to work on any Rust project for a very long time. My apologies for leaving the code half-baked :(

tertsdiepraam commented:
I tried to salvage this PR, but I couldn't get it quite right, so I'm closing it. Feel free to open a new one when you want to work on it again, starting from the current master branch!
